Query Builder

sahil143 opened the pull request:
Smart Query Builder Story: https://openshift.io/openshiftio/Openshift_io/plan/detail/383

sanbornsen committed:
fix(query): string enclouser added to query values

- an optional quotation enclosure is added for the values in query
- area.name: "new area" and area.name: new area are same

sanbornsen committed:
feat(query): query suggestion service added

sanbornsen committed:
feat(query): query suggestion added to the component

sanbornsen committed:
fix(query): iteration added to the suggestion

sahil143 committed:
fix(query): remove suggestions on blur

sanbornsen committed:
fix(query): sugestion replace function added

sahil143 committed:
fix(query): add dropdown for suggestions

sahil143 committed:
fix(query): add delay to query suggestion

sanbornsen committed:
fix(query): suggestion on click replace the text added

sahil143 committed:
fix(query): dont close dropdown on blur

sahil143 committed:
fix(query): fix replace suggestions

sahil143 committed:
fix(query): add css for dropdown-item

sanbornsen committed:
fix(query): new fields added to the suggestion

sahil143 committed:
fix(angular/cdk): add angular cdk to package.json

sahil143 committed:
fix(list): add dropdown list component for Highlight interface from cdk

sahil143 committed:
fix(query): add a11y using ActiveDescendantKeyManager on suggestion dropdown

sanbornsen committed:
fix(query): fixed unit test for query

sahil143 commented:
[test]

sahil143 committed:
fix(query): add support for left and right arrow keys to suggest query fields

sanbornsen committed:
fix(less): generic class fixed

Raunak1203 commented:
[test]

christianvogt commented:
@sahil143 is this ready for review or are there still changes being made?

sanbornsen commented:
[test]

sahil143 committed:
fix(tests): fix failing  functional tests

sanbornsen reviewed (approved):
Looks good to me

sahil143 commented:
[test]

christianvogt commented:
@sahil143 I get this error whenever the query is incorrect. It's very intrusive. An error message below the input would probably be better.
![image](https://user-images.githubusercontent.com/14068621/49403620-3c247500-f71b-11e8-8780-d0c371cdfac1.png)

That being said, when the query is correct it seems to work as expected.
There's also some oddities with the content assist when the cursor isn't at the end of the input.

christianvogt commented:
When I press enter and the query completes, focus is lost from the input. I would expect focus to remain in the query input field.

christianvogt commented:
Sometimes pressing enter executes the query and sometimes it performs an auto complete. It's confusing and often messes up the query. Worst part is that after the query gets messed up I cannot simply press cmd+z to undo, because it doesn't undo the auto complete entry.

christianvogt commented:
Got the following error when clicking on a query item to go to the details and then pressing the back button:

```
ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'ngIf: false'. Current value: 'ngIf: true'.
    at viewDebugError (VM581881 core.js:17215)
    at expressionChangedAfterItHasBeenCheckedError (VM581881 core.js:17203)
    at checkBindingNoChanges (VM581881 core.js:17305)
    at checkNoChangesNodeInline (VM581881 core.js:20176)
    at checkNoChangesNode (VM581881 core.js:20165)
    at debugCheckNoChangesNode (VM581881 core.js:20769)
    at debugCheckDirectivesFn (VM581881 core.js:20697)
    at Object.eval [as updateDirectives] (PlannerQueryComponent.html:22)
    at Object.debugUpdateDirectives [as updateDirectives] (VM581881 core.js:20686)
    at checkNoChangesView (VM581881 core.js:20064)
    at callViewAction (VM581881 core.js:20303)
    at execComponentViewsAction (VM581881 core.js:20265)
    at checkNoChangesView (VM581881 core.js:20067)
    at callViewAction (VM581881 core.js:20303)
    at execEmbeddedViewsAction (VM581881 core.js:20286)
    at checkNoChangesView (VM581881 core.js:20065)
    at callViewAction (VM581881 core.js:20303)
    at execComponentViewsAction (VM581881 core.js:20265)
    at checkNoChangesView (VM581881 core.js:20067)
    at callWithDebugContext (VM581881 core.js:20976)
    at Object.debugCheckNoChangesView [as checkNoChangesView] (VM581881 core.js:20657)
    at ViewRef_.checkNoChanges (VM581881 core.js:18471)
    at eval (VM581881 core.js:15127)
    at Array.forEach (<anonymous>)
    at ApplicationRef.tick (VM581881 core.js:15127)
    at eval (VM581881 core.js:15016)
    at ZoneDelegate.invoke (zone.js?f146:388)
    at Object.onInvoke (VM581881 core.js:14397)
    at ZoneDelegate.invoke (zone.js?f146:387)
    at Zone.run (zone.js?f146:138)
    at NgZone.run (VM581881 core.js:14311)
    at Object.next (VM581881 core.js:15016)
    at SafeSubscriber.schedulerFn [as _next] (VM581881 core.js:10575)
    at SafeSubscriber.__tryOrUnsub (VM581886 Subscriber.js:206)
    at SafeSubscriber.next (VM581886 Subscriber.js:144)
    at Subscriber._next (VM581886 Subscriber.js:87)
    at Subscriber.next (VM581886 Subscriber.js:64)
    at EventEmitter.Subject.next (VM581903 Subject.js:58)
    at EventEmitter.emit (VM581881 core.js:10559)
    at checkStable (VM581881 core.js:14366)
    at Object.onHasTask (VM581881 core.js:14410)
    at ZoneDelegate.hasTask (zone.js?f146:441)
    at ZoneDelegate._updateTaskCount (zone.js?f146:461)
    at Zone._updateTaskCount (zone.js?f146:285)
    at Zone.runTask (zone.js?f146:205)
    at drainMicroTaskQueue (zone.js?f146:595)
    at ZoneTask.invokeTask [as invoke] (zone.js?f146:500)
    at invokeTask (zone.js?f146:1540)
    at XMLHttpRequest.globalZoneAwareCallback (zone.js?f146:1566)
console.(anonymous function) @ console.js?7ccc:35
View_PlannerQueryComponent_0 @ PlannerQueryComponent.html:22
proxyClass @ compiler.js?ad3e:17129
DebugContext_.logError @ core.js?09c9:20601
ErrorHandler.handleError @ core.js?09c9:12549
RavenExceptionHandler.handleError @ exception.handler.ts?ec0d:26
(anonymous) @ core.js?09c9:14795
ZoneDelegate.invoke @ zone.js?f146:388
Zone.run @ zone.js?f146:138
NgZone.runOutsideAngular @ core.js?09c9:14019
ApplicationRef.tick @ core.js?09c9:14795
(anonymous) @ core.js?09c9:14679
ZoneDelegate.invoke @ zone.js?f146:388
onInvoke @ core.js?09c9:14060
ZoneDelegate.invoke @ zone.js?f146:387
Zone.run @ zone.js?f146:138
NgZone.run @ core.js?09c9:13974
next @ core.js?09c9:14679
schedulerFn @ core.js?09c9:10238
SafeSubscriber.__tryOrUnsub @ Subscriber.js?1453:196
SafeSubscriber.next @ Subscriber.js?1453:134
Subscriber._next @ Subscriber.js?1453:77
Subscriber.next @ Subscriber.js?1453:54
Subject.next @ Subject.js?2bd2:47
EventEmitter.emit @ core.js?09c9:10222
checkStable @ core.js?09c9:14029
onHasTask @ core.js?09c9:14073
ZoneDelegate.hasTask @ zone.js?f146:441
ZoneDelegate._updateTaskCount @ zone.js?f146:461
Zone._updateTaskCount @ zone.js?f146:285
Zone.runTask @ zone.js?f146:205
drainMicroTaskQueue @ zone.js?f146:595
ZoneTask.invokeTask @ zone.js?f146:500
invokeTask @ zone.js?f146:1540
globalZoneAwareCallback @ zone.js?f146:1566
error (async)
customScheduleGlobal @ zone.js?f146:1666
ZoneDelegate.scheduleTask @ zone.js?f146:407
onScheduleTask @ long-stack-trace-zone.js?8123:116
ZoneDelegate.scheduleTask @ zone.js?f146:401
onScheduleTask @ zone.js?f146:297
ZoneDelegate.scheduleTask @ zone.js?f146:401
Zone.scheduleTask @ zone.js?f146:232
Zone.scheduleEventTask @ zone.js?f146:258
(anonymous) @ zone.js?f146:1831
(anonymous) @ raven.js?72aa:1202
(anonymous) @ http.js?b7f3:1619
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
(anonymous) @ cache.interceptor.ts?da6b:37
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
DoOperator.call @ tap.js?c4cc:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
(anonymous) @ scalar.js?3d4f:5
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
FilterOperator.call @ filter.js?5670:15
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
CatchOperator.call @ catchError.js?f59d:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
CatchOperator.call @ catchError.js?f59d:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
(anonymous) @ subscribeToArray.js?fad2:5
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
ScanOperator.call @ scan.js?742d:23
Observable.subscribe @ Observable.js?e9b9:24
TakeLastOperator.call @ takeLast.js?0318:24
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
TapSubscriber._next @ tap.js?c4cc:51
Subscriber.next @ Subscriber.js?1453:54
(anonymous) @ scalar.js?3d4f:5
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
DoOperator.call @ tap.js?c4cc:18
Observable.subscribe @ Observable.js?e9b9:24
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
DoOperator.call @ tap.js?c4cc:18
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
SwitchMapSubscriber._innerSub @ switchMap.js?d792:51
SwitchMapSubscriber._next @ switchMap.js?d792:41
Subscriber.next @ Subscriber.js?1453:54
FilterSubscriber._next @ filter.js?5670:38
Subscriber.next @ Subscriber.js?1453:54
TapSubscriber._next @ tap.js?c4cc:51
Subscriber.next @ Subscriber.js?1453:54
MergeMapSubscriber.notifyNext @ mergeMap.js?a6c5:84
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
MergeMapSubscriber.notifyNext @ mergeMap.js?a6c5:84
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
EverySubscriber.notifyComplete @ every.js?cc06:30
EverySubscriber._complete @ every.js?cc06:47
Subscriber.complete @ Subscriber.js?1453:66
MergeMapSubscriber._complete @ mergeMap.js?a6c5:79
Subscriber.complete @ Subscriber.js?1453:66
(anonymous) @ subscribeToArray.js?fad2:8
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
EveryOperator.call @ every.js?cc06:14
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
EverySubscriber.notifyComplete @ every.js?cc06:30
EverySubscriber._complete @ every.js?cc06:47
Subscriber.complete @ Subscriber.js?1453:66
MergeMapSubscriber._complete @ mergeMap.js?a6c5:79
Subscriber.complete @ Subscriber.js?1453:66
(anonymous) @ subscribeToArray.js?fad2:8
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
EveryOperator.call @ every.js?cc06:14
Observable.subscribe @ Observable.js?e9b9:24
MergeMapOperator.call @ mergeMap.js?a6c5:29
Observable.subscribe @ Observable.js?e9b9:24
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
TapSubscriber._next @ tap.js?c4cc:51
Subscriber.next @ Subscriber.js?1453:54
SwitchMapSubscriber.notifyNext @ switchMap.js?d792:72
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
(anonymous) @ scalar.js?3d4f:5
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
SwitchMapSubscriber._innerSub @ switchMap.js?d792:51
SwitchMapSubscriber._next @ switchMap.js?d792:41
Subscriber.next @ Subscriber.js?1453:54
SwitchMapSubscriber.notifyNext @ switchMap.js?d792:72
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
TapSubscriber._next @ tap.js?c4cc:51
Subscriber.next @ Subscriber.js?1453:54
MergeMapSubscriber.notifyNext @ mergeMap.js?a6c5:84
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
(anonymous) @ scalar.js?3d4f:5
Observable._trySubscribe @ Observable.js?e9b9:43
Observable.subscribe @ Observable.js?e9b9:29
MapOperator.call @ map.js?ebb6:18
Observable.subscribe @ Observable.js?e9b9:24
(anonymous) @ subscribeTo.js?1716:21
subscribeToResult @ subscribeToResult.js?ce8b:11
MergeMapSubscriber._innerSub @ mergeMap.js?a6c5:74
MergeMapSubscriber._tryNext @ mergeMap.js?a6c5:68
MergeMapSubscriber._next @ mergeMap.js?a6c5:51
Subscriber.next @ Subscriber.js?1453:54
SwitchMapSubscriber.notifyNext @ switchMap.js?d792:72
InnerSubscriber._next @ InnerSubscriber.js?acf8:15
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
Subscriber._next @ Subscriber.js?1453:77
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
MapSubscriber._next @ map.js?ebb6:41
Subscriber.next @ Subscriber.js?1453:54
TapSubscriber._next @ tap.js?c4cc:51
Subscriber.next @ Subscriber.js?1453:54
TakeLastSubscriber._complete @ takeLast.js?0318:57
Subscriber.complete @ Subscriber.js?1453:66
MergeMapSubscriber._complete @ mergeMap.js?a6c5:79
Subscriber.complete @ Subscriber.js?1453:66
(anonymous) @ scalar.js?3d4f:6
setTimeout (async)
scheduleTask @ zone.js?f146:2075
ZoneDelegate.scheduleTask @ zone.js?f146:407
onScheduleTask @ long-stack-trace-zone.js?8123:116
ZoneDelegate.scheduleTask @ zone.js?f146:401
onScheduleTask @ zone.js?f146:297
ZoneDelegate.scheduleTask @ zone.js?f146:401
Zone.scheduleTask @ zone.js?f146:232
Zone.scheduleMacroTask @ zone.js?f146:255
scheduleMacroTaskWithCurrentZone @ zone.js?f146:1114
(anonymous) @ zone.js?f146:2090
proto.(anonymous function) @ zone.js?f146:1394
(anonymous) @ raven.js?72aa:1133
(anonymous) @ router.js?6580:3909
schedulerFn @ core.js?09c9:10238
SafeSubscriber.__tryOrUnsub @ Subscriber.js?1453:196
SafeSubscriber.next @ Subscriber.js?1453:134
Subscriber._next @ Subscriber.js?1453:77
Subscriber.next @ Subscriber.js?1453:54
Subject.next @ Subject.js?2bd2:47
EventEmitter.emit @ core.js?09c9:10222
(anonymous) @ common.js?229d:151
wrapped @ raven.js?72aa:376
ZoneDelegate.invokeTask @ zone.js?f146:421
onInvokeTask @ core.js?09c9:14051
ZoneDelegate.invokeTask @ zone.js?f146:420
Zone.runTask @ zone.js?f146:188
ZoneTask.invokeTask @ zone.js?f146:496
invokeTask @ zone.js?f146:1540
globalZoneAwareCallback @ zone.js?f146:1577
```

sahil143 commented:
>Got the following error when clicking on a query item to go to the details and then pressing the back button:
>ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'ngIf: false'. Current value: 'ngIf: true'.
>View_PlannerQueryComponent_0 | @ | PlannerQueryComponent.html:22

Fixed in latest commit

>Sometimes pressing enter executes the query and sometimes it performs an auto complete

When suggestion dropdown is open pressing enter auto completes the query and when dropdown is close pressing enter executes the query

>Worst part is that after the query gets messed up I cannot simply press cmd+z to undo, because it doesn't undo the auto complete entry.

Solution for this would be to keep our own history(ctrl z) for query input field?

> When I press enter and the query completes, focus is lost from the input. I would expect focus to remain in the query input field.

This is intended behaviour but I dont see why we can't focus on input after pressing enter. Fixed in latest commit

> I get this error whenever the query is incorrect. It's very intrusive. An error message below the input would probably be better.

We dont have query validator which can validate the query. We send the api request for invalid queries and when backend throws error we show this notification error. Also, Area below input has other text. Adding error below input field would disrupt the experience. WDYT?

@christianvogt ^^

sahil143 committed:
fix(query): fix `expression change error`, focus on input after pressing enter

joshuawilson commented:
@christianvogt is this ready to be merged?

christianvogt commented:
@sahil143 there's something wrong w/ the auto complete and the enter key. After entering in the value, pressing enter again replaced my value with the field again. I would expect the query to be issued at this point when pressing enter.
![query-enter](https://user-images.githubusercontent.com/14068621/49657279-51481f00-fa0d-11e8-875e-eeeff9b4138c.gif)

christianvogt commented on src/app/components_ngrx/planner-query/planner-query.component.html:
```diff
@@ -15,18 +15,37 @@ <h4>Please wait, we are loading your data.</h4>
+			<span
+				*ngIf="searchQuery !== '' && !queryInput.disabled"
+				class="close f8-query__clear-button"
+				[class.pficon-spinner2]="valueLoading"
```
I'm not sure this adds much value. It continuously blinks while typing. Considering the suggestions is an added value that appears automatically, i don't think we need to show a spinner while it's loading?
We can keep it, I'd have UXD comment on this behavior.

sahil143 replied:
We added the spinner here because for query `assignee: USER_NAME` we make a backend request to fetch the users and it takes some time.

christianvogt replied:
But it is happening on every keystroke which is a bit excessive. IIRC i inspected the network and saw no requests being made while the spinner was blinking as I typed.

sahil143 replied:
I can't reproduce this one. In my Case, request is being made to backend

christianvogt commented on src/app/components_ngrx/planner-query/planner-query.component.ts:
```diff
@@ -163,30 +189,98 @@ export class PlannerQueryComponent implements OnInit, OnDestroy, AfterViewChecke
-        });
+    // If Enter pressed
+    if (this.isSuggestionDropdownOpen) {
+      if (keycode === 13) {
```
Just because the menu is opened, doesn't mean we should add a suggestion on enter. It should only do it if there is a menu item selected.

christianvogt commented on src/app/services/filter.service.ts:
```diff
@@ -194,6 +201,40 @@ export class FilterService {
+    // and there no enclouser already
+    // then enclosed the string and return it
+    if (
+      query.split(' ').length > 1 &&
```
Better to check for `indexOf` rather than splitting to create an array.

christianvogt commented on src/app/services/filter.service.ts:
```diff
@@ -6,20 +6,27 @@ import { FilterModel } from '../models/filter.model';
+  public in_notation = IN;
+  public not_in_notation = NOT_IN;
+  public sub_str_notation = SUB_STR;
+  public str_enclouser = ENCLOUSER;
```
What's the point of assigning these instance values from constants? They don't change, and shouldn't. The code refers to both the constants and the instance vars. You should be able to clean these up.

christianvogt commented on src/app/services/filter.service.ts:
```diff
@@ -194,6 +201,40 @@ export class FilterService {
+   * @param query
+   */
+  clearEnclosedValue(query: string) {
+    if (query[0] == this.str_enclouser && query[query.length - 1] == this.str_enclouser) {
```
Should use `===` everywhere. There are more occurrences of this elsewhere.

christianvogt commented on src/app/widgets/list-item/list-item.component.less:
```diff
@@ -0,0 +1,16 @@
+@import (reference) '../../../assets/stylesheets/_base';
+
+.dropdown-item {
```
This class will affect all `dropdown-item`s in the app.

christianvogt reviewed (commented):
I get this error when refreshing the query page w/ a query present. Then going to the input field and immediately pressing enter.
```
ERROR TypeError: Cannot read property 'item' of undefined
    at PlannerQueryComponent.fetchWorkItemForQuery (planner-query.component.ts?6c46:201)
    at Object.eval [as handleEvent] (PlannerQueryComponent.html:33)
```

sahil143 committed:
fix(query): remove unused imports and refactor code

sahil143 committed:
fix(query): clear the activeItem from dropdown on Enter

sahil143 committed:
chore(refactor): use constants from query-keys instead of public instances from filterService

joshuawilson committed:
Merge branch 'master' into query
